libraries/anvill_passes/src/TransformRemillJumpIntrinsics.cpp

```diff
@@ -0,0 +1,229 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <anvill/Analysis/CrossReferenceResolver.h>
+#include <anvill/Analysis/Utils.h>
+#include <anvill/Lifters/EntityLifter.h>
+#include <anvill/Transforms.h>
+#include <glog/logging.h>
+#include <llvm/ADT/Triple.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/InstIterator.h>
+#include <llvm/IR/Instruction.h>
+#include <llvm/IR/Intrinsics.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Pass.h>
+#include <remill/BC/Util.h>
+
+#include <utility>
+#include <vector>
+
+
+namespace anvill {
+namespace {
+
+const std::string_view remill_jump = "__remill_jump";
+const std::string_view remill_function_return = "__remill_function_return";
+
+enum ReturnAddressResult {
+
+  // This is a case where a value returned by `llvm.returnaddress`, or
+  // casted from `__anvill_ra`, reaches into the `pc` argument of the
+  // `__remill_jump` intrinsic. This is the ideal case that we want to
+  // replace it with `__remill_function_return`.
+  kReturnAddressProgramCounter,
+
+  // This is a case a value returned by `llvm.returnaddress`, or casted
+  // from `__anvill_ra` does not reaches to the `pc` argument and it
+  // should not get transformed to `__remill_function_return`.
+  kUnclassifiableProgramCounter
+};
+
+class TransformRemillJumpIntrinsics final : public llvm::FunctionPass {
+ public:
+  TransformRemillJumpIntrinsics(const EntityLifter &lifter_)
+      : llvm::FunctionPass(ID),
+        xref_resolver_(lifter_) {}
+
+  bool runOnFunction(llvm::Function &func) final;
+
+ private:
+  ReturnAddressResult QueryReturnAddress(const llvm::DataLayout &dl,
+                                         llvm::Value *val) const;
+
+  bool TransformJumpIntrinsic(llvm::CallBase *call);
+
+  static char ID;
+  const CrossReferenceResolver xref_resolver_;
+};
+
+
+char TransformRemillJumpIntrinsics::ID = '\0';
+
+// Returns `true` if `val` is a possible return address
+ReturnAddressResult
+TransformRemillJumpIntrinsics::QueryReturnAddress(const llvm::DataLayout &dl,
+                                                  llvm::Value *val) const {
+
+  if (auto call = llvm::dyn_cast<llvm::CallBase>(val)) {
+    if (call->getIntrinsicID() == llvm::Intrinsic::returnaddress) {
+      return kReturnAddressProgramCounter;
+    } else if (auto func = call->getCalledFunction()) {
```

Comment:
This is an interesting case that may be missing from the aforementioned functions, i.e. a remill memory read (and missing, a load inst case. Perhaps update the analysis function with these cases you've marked.